validator_client/src/config.rs

```diff
@@ -1,32 +1,131 @@
+use bincode;
+use bls::Keypair;
+use clap::ArgMatches;
 use std::fs;
+use std::fs::File;
+use std::io::{Error, ErrorKind};
 use std::path::PathBuf;
 use types::ChainSpec;
 
 /// Stores the core configuration for this validator instance.
 #[derive(Clone)]
-pub struct ClientConfig {
+pub struct ValidatorClientConfig {
+    /// The data directory, which stores all validator databases
     pub data_dir: PathBuf,
+    /// The directory where the individual validator configuration directories are stored.
+    pub validator_dir: PathBuf,
+    /// The server at which the Beacon Node can be contacted
     pub server: String,
+    /// The chain specification that we are connecting to
     pub spec: ChainSpec,
 }
 
-const DEFAULT_LIGHTHOUSE_DIR: &str = ".lighthouse-validators";
+const DEFAULT_VALIDATOR_DATADIR: &str = ".lighthouse-validator";
+const DEFAULT_VALIDATORS_SUBDIR: &str = "validators";
+const DEFAULT_PRIVATE_KEY_FILENAME: &str = "private.key";
 
-impl ClientConfig {
-    /// Build a new configuration from defaults.
-    pub fn default() -> Self {
-        let data_dir = {
-            let home = dirs::home_dir().expect("Unable to determine home dir.");
-            home.join(DEFAULT_LIGHTHOUSE_DIR)
+impl ValidatorClientConfig {
+    /// Build a new configuration from defaults, which are overrided by arguments provided.
+    pub fn build_config(arguments: &ArgMatches) -> Result<Self, Error> {
```

Comment:
It is conventional for functions on structs that return `Self` to be called `new`. Perhaps you could implement `new()` which gives `default()` values and have another function `parse_args()` which performs the argument passing.

Reply:
Here is a version of a config and how I did it:
https://github.com/sigp/lighthouse/blob/network-server/beacon_node/client/src/client_config.rs

Reply:
(This is one of the configs i plan to rename :p)

Reply:
In my example, i don't implement `new()` but I do implement `default()`.

Reply:
Yeah nice one.